### PR TITLE
staging/service: refactor parser for sysvinit/systemd result

### PR DESCRIPTION
**Pull request by PandaWei**
we should obey the rule that one function do one thing at a time,and do well.
so...

make it more tidy, readable and easier to understand.
modifications are mainly concentrated in the following,
* sys_v_init_result_parser
* systemd_result_parser

### CODE REVIEW

**Comment by yangdongsheng** [comment]
Please list more detail about your cleanup patch here.

**Comment by yangdongsheng** [comment]
From a simple look at it, it seems breaking the OOB design. Sounds a bit strange to me.

**Comment by PandaWei** [comment]
Forgive that I have no knowledge of OOB. ^_^
do you mean OOD?

**Comment by yangdongsheng** [comment]
yea, my bad. OOB is another glossary in mtd driver.

**Comment by PandaWei** [comment]
fc22ce7:
staging/service: extract the common as a super-class
    
    * add new super class named _ServiceCommTool that extract the
      common from _ServiceResultParser and _ServiceCommandGenerator.
      and make the codes more clearly by inheritance.
    * polish the docstring and comments
---------------------------------------------------------------------------------------
67dee39
staging/service: refactor parser for sysvinit/systemd
    
    The parser function looks very massive and messy, and
    hasn't an unclean structure.
    If we want to add new feature to check other kind of CmdResult,
    There's only one way, just to add new else-block statement.
    Just like d92e213,
    
    One function do one thing at a time and do well,
    which is a golden rule.
    so it's necessary to refactor them and make it tidy,
    readable and easier to understand...
    
    Modifications includes the following,
    * sys_v_init_result_parser()
    * systemd_result_parser()
    * polish the docstring and comments
    
    Just take systemd_result_parser() as an example,
    Split sys_v_init_result_parser into smaller functions,
    * sys_v_init_result_parser
      |- sysd_parse_status (systemctl status)
      |- sysv_parse_list   (systemctl list)
      |- parse_raw_status  (shared by systemd and sysvinit)
      |- _ServiceResultParser.default_method (default)
    
    This commit allow people add new feature easily,
    As follow,
    1) Just coding the new parser function at first
    2) and then, declare this function at parser-dict[]

**Comment by PandaWei** [comment]
@yangdongsheng 
Hi, Update it as your demand.
Please help to check it. 
Thanks in advance.

File: virttest/staging/service.py [hunk 1]
```diff
@@ -88,184 +88,163 @@
...
+
+def sys_v_init_result_parser(command):
+    """
+    Parse results for sys_v style commands.
+
+    :param command: service sub-command(string).
+    :return: depends on sub-command.
+    """
+    if command in _sysv_parser_dict.keys():
+        return _sysv_parser_dict.get(command)
```
Comment (yangdongsheng) [inline virttest/staging/service.py]: At first, it's a good idea to do this cleanup. Secondly, there is a problem that you changed these name from parser to parse_xxx. In this function, it's init_parser() but you are returning dict.get(command) which is a process named as  parse_xxx(). Then I suggest you change the parse_xxx() to xx_parser(). Third, in command generator part, there is a similar problem we need a lot of if/else. But that's not as bad as the parser part. If you think it's worth to do, I suggest you to do a similar cleanup for the command_generator part. At last, you are using sysv and sysd as the prefix to distinguish the different os. I think it's okey, but I think you'd better change the all prefix as this standard if you like this style.

Comment (PandaWei) [inline virttest/staging/service.py]: Thanks for such a detailed explanation and comment.

**Comment by PandaWei** [comment]
Updated.
the class declared and started by SysVInit/Systemd
SO I adopt the  SysVInit/Systemd  as the prefix(s) for parsers.

In order to keep uniform format,
I change the prefix of command generator function and submit a new commit.

As you said, command generator part is *not* as bad as the parser part.
and A bug introduced by commit:"Add option m to the tar command" in command generator part.
so I decided don't modify the command generator funciton this time.
if the bug is fixed, i will continue this work.

**Comment by PandaWei** [comment]
The CI failed. it's abnormal from Monday.
so I attach the result of local check.
[root@localhost staging]# inspekt style service.py
PEP8 compliance check PASS
[root@localhost staging]# inspekt lint service.py
Syntax check PASS
[root@localhost staging]# inspekt indent service.py
Indentation check PASS

**Comment by yangdongsheng** [comment]
A bug was introduced into service module? Then please wait for this bug fixed. what kind of the bug introduced by "Add option m to the tar command"?  I can't imagine the relationship between this commit with service module.

**Comment by yangdongsheng** [comment]
Don't worry, this is not a urgent PR, let's wait for the CI fixed.

**Comment by PandaWei** [comment]
Bug is not inaccurate.
Maybe we can regard it's a redundance.
and the author said he will submit a new PR to fix it.

The detail mentioned in ML, as follow.
@yangdongsheng
Hi, yes I agree, the command_name = "service" is defined above already and command_name = "chkconfig" is not needed .
I will remove that line. Shell I request a new pull request?

The variable of command_name declared and used by command generator part

**Comment by waynesun09** [comment]
@PandaWei 

In travis CI log:
1) /home/travis/build/autotest/virt-test/virttest/staging/service.py:docstring of virttest.staging.service.SysVInit_list_parser:7: ERROR: Unexpected indentation.
2) /home/travis/build/autotest/virt-test/virttest/staging/service.py:docstring of virttest.staging.service.SysVInit_list_parser:10: WARNING: Block quote ends without a blank line; unexpected unindent.

**Comment by yangdongsheng** [comment]
Okey, so it was not introduced by "Add option m to the tar command", instead, it's related with the patch of "Added new method statuslog() to show the service status text" from @svirt. Okey, I see it now. Good, let's wait for this problem fixed by another PR from @svirt.

**Comment by yangdongsheng** [comment]
ADDITION: As this is a utils module used by lots of other modules or testcases, please do a regression test before we want these patch in.

**Comment by PandaWei** [comment]
@waynesun09
Thanks for your reminder.
I will check and update it.

**Comment by PandaWei** [comment]
rebase it.
and add unitest for service module.

File: virttest/service_unittest.py [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+
+class FactoryHelperTest(ModuleLoad):
+
+    def test_dict_attribute(self):
+        self.assertTrue(hasattr(self.service.Factory.FactoryHelper,
+                                '_command_generators'))
+        self.assertTrue(hasattr(self.service.Factory.FactoryHelper,
+                                '_result_parsers'))
+        self.assertTrue(hasattr(self.service.Factory.FactoryHelper,
+                                '_service_managers'))
```
Comment (yangdongsheng) [inline virttest/service_unittest.py]: Although, adding a unittest for service module is a good idea, but these checks do not make sense to me. Why we need to make sure a FactoryHelper has a attribute named as _result_parsers?? that's not a rule and It's very possible to return FALSE if we reimplemented this module in a better way. What we have to test in a unittest is the logic and api for a module, do not go too deeply in it.

File: virttest/service_unittest.py [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+        del factory
+
+    def test_create_generic_service(self):
+        generic_instance = self.service.Factory.create_generic_service()
+        self.assertTrue(isinstance(generic_instance,
+                                   self.service._GenericServiceManager))
+        del generic_instance
+
+    def test_create_specific_service(self):
+        specific_instance = self.service.Factory.create_specific_service("network")
```
Comment (yangdongsheng) [inline virttest/service_unittest.py]: It's not a good idea to make a unittest depending on the environment. Please think about that, there is no service named as "network", then this unittest will fail. Please create a fake service for it if necessary. `github-autotest/github-autotest/client/shared/test_utils/mock.py` is a good tool for that. And you can refer to the service_unittest.py in `github-autotest/github-autotest/client/shared`.

**Comment by PandaWei** [comment]
Update in accordance with comments.

**Comment by will-Do** [comment]
The series patches are good, just two problems from my review:
1. Function name should use lower case all the time.
2. A typo in you patch 4 summary(s/unitest/unittest)

**Comment by PandaWei** [comment]
Thanks @will-Do 
I will update it as your comments.

**Comment by PandaWei** [comment]
Updated.
declare function using lower case.

**Comment by yangdongsheng** [comment]
Okey, the last question, what's the result of your unittest and other tests related with service module? Please post them on this PR then we can merge it.

**Comment by PandaWei** [comment]
[root@localhost virttest]# python service_unittest.py 
..............
----------------------------------------------------------------------
Ran 14 tests in 0.116s
OK

**Comment by yangdongsheng** [comment]
What about the other *real* cases using service module?

**Comment by PandaWei** [comment]
OK, just take virsh.domstate as example.
These cases will call libvirtd service, for start /stop/restart.
and just for your information.
The __init__ of class Libvirtd will call service.Factory.create_service()

[root@localhost virt-test]# ./run -t libvirt --no-downloads -k --keep-image-between-tests --test "virsh.domstate"
SETUP: PASS (0.70 s)
DATA DIR: /var/lib/virt_test
DEBUG LOG: /root/osc/virt-test/logs/run-2015-04-16-14.19.09/debug.log
TESTS: 22
(1/22) type_specific.io-github-autotest-libvirt.virsh.domstate.normal_test.id_option: PASS (1.64 s)
(2/22) type_specific.io-github-autotest-libvirt.virsh.domstate.normal_test.name_option: PASS (1.37 s)
(3/22) type_specific.io-github-autotest-libvirt.virsh.domstate.normal_test.paused_option: PASS (1.47 s)
(4/22) type_specific.io-github-autotest-libvirt.virsh.domstate.normal_test.uuid_option: PASS (1.36 s)
(5/22) type_specific.io-github-autotest-libvirt.virsh.domstate.normal_test.remote: SKIP
(6/22) type_specific.io-github-autotest-libvirt.virsh.domstate.normal_test.reason.suspend_vm: PASS (1.38 s)
(7/22) type_specific.io-github-autotest-libvirt.virsh.domstate.normal_test.reason.resume_vm: PASS (1.40 s)
(8/22) type_specific.io-github-autotest-libvirt.virsh.domstate.normal_test.reason.destroy_vm: PASS (1.36 s)
(9/22) type_specific.io-github-autotest-libvirt.virsh.domstate.normal_test.reason.start_vm: PASS (1.87 s)
(10/22) type_specific.io-github-autotest-libvirt.virsh.domstate.normal_test.reason.crash_vm.oncrash_destroy: SKIP
(11/22) type_specific.io-github-autotest-libvirt.virsh.domstate.normal_test.reason.crash_vm.oncrash_restart: SKIP
(12/22) type_specific.io-github-autotest-libvirt.virsh.domstate.normal_test.reason.crash_vm.oncrash_preserve: SKIP
(13/22) type_specific.io-github-autotest-libvirt.virsh.domstate.normal_test.reason.crash_vm.oncrash_coredump_destroy: SKIP
(14/22) type_specific.io-github-autotest-libvirt.virsh.domstate.normal_test.reason.crash_vm.oncrash_coredump_restart: SKIP
(15/22) type_specific.io-github-autotest-libvirt.virsh.domstate.error_test.no_option: PASS (1.37 s)
(16/22) type_specific.io-github-autotest-libvirt.virsh.domstate.error_test.hex_id_option: PASS (1.68 s)
(17/22) type_specific.io-github-autotest-libvirt.virsh.domstate.error_test.invalid_id_option: PASS (1.39 s)
(18/22) type_specific.io-github-autotest-libvirt.virsh.domstate.error_test.invalid_uuid_option: PASS (1.40 s)
(19/22) type_specific.io-github-autotest-libvirt.virsh.domstate.error_test.invalid_name_option: PASS (1.41 s)
(20/22) type_specific.io-github-autotest-libvirt.virsh.domstate.error_test.extra_option: PASS (1.38 s)
(21/22) type_specific.io-github-autotest-libvirt.virsh.domstate.error_test.with_libvirt_stop: PASS (3.39 s)
(22/22) type_specific.io-github-autotest-libvirt.virsh.domstate.error_test.kill_vm_after_stopping_libvirtd: PASS (3.15 s)
TOTAL TIME: 49.19 s
TESTS PASSED: 16
TESTS FAILED: 0
TESTS SKIPPED: 6
SUCCESS RATE: 100.00 %

**Comment by yangdongsheng** [comment]
Would you mind do a more regression test on it? What about the other *real* cases using service? And what about environment of systemd?

**Comment by PandaWei** [comment]
OK, later,
I will post the result for other service related cases

**Comment by PandaWei** [comment]
This RT for cgroup service .

./run -t libvirt --no-downloads -k --keep-image-between-tests --tests 'virsh.numatune'
SETUP: PASS (0.80 s)
DATA DIR: /var/lib/virt_test
DEBUG LOG: /root/osc/virt-test/logs/run-2015-04-16-18.23.38/debug.log
TESTS: 60
(1/60) type_specific.io-github-autotest-libvirt.virsh.numatune.positive_testing.set_numa_parameter.shutoff_guest.change_mode.options.config.preferred: PASS (2.90 s)
(2/60) type_specific.io-github-autotest-libvirt.virsh.numatune.positive_testing.set_numa_parameter.shutoff_guest.change_mode.options.config.interleave: PASS (1.30 s)
(3/60) type_specific.io-github-autotest-libvirt.virsh.numatune.positive_testing.set_numa_parameter.shutoff_guest.change_mode.options.config.strict: PASS (1.29 s)
(4/60) type_specific.io-github-autotest-libvirt.virsh.numatune.positive_testing.set_numa_parameter.shutoff_guest.change_mode.options.current.preferred: PASS (1.30 s)
(5/60) type_specific.io-github-autotest-libvirt.virsh.numatune.positive_testing.set_numa_parameter.shutoff_guest.change_mode.options.current.interleave: PASS (1.28 s)
(6/60) type_specific.io-github-autotest-libvirt.virsh.numatune.positive_testing.set_numa_parameter.shutoff_guest.change_mode.options.current.strict: PASS (1.28 s)
(7/60) type_specific.io-github-autotest-libvirt.virsh.numatune.positive_testing.set_numa_parameter.shutoff_guest.change_nodeset.options.config.comma_list: PASS (1.06 s)
(8/60) type_specific.io-github-autotest-libvirt.virsh.numatune.positive_testing.set_numa_parameter.shutoff_guest.change_nodeset.options.config.ranges: PASS (0.87 s)
(9/60) type_specific.io-github-autotest-libvirt.virsh.numatune.positive_testing.set_numa_parameter.shutoff_guest.change_nodeset.options.config.excluding: PASS (0.83 s)
(10/60) type_specific.io-github-autotest-libvirt.virsh.numatune.positive_testing.set_numa_parameter.shutoff_guest.change_nodeset.options.config.single: PASS (1.30 s)
(11/60) type_specific.io-github-autotest-libvirt.virsh.numatune.positive_testing.set_numa_parameter.shutoff_guest.change_nodeset.options.current.comma_list: PASS (1.04 s)
(12/60) type_specific.io-github-autotest-libvirt.virsh.numatune.positive_testing.set_numa_parameter.shutoff_guest.change_nodeset.options.current.ranges: PASS (0.81 s)
(13/60) type_specific.io-github-autotest-libvirt.virsh.numatune.positive_testing.set_numa_parameter.shutoff_guest.change_nodeset.options.current.excluding: PASS (0.82 s)
(14/60) type_specific.io-github-autotest-libvirt.virsh.numatune.positive_testing.set_numa_parameter.shutoff_guest.change_nodeset.options.current.single: PASS (1.31 s)
(15/60) type_specific.io-github-autotest-libvirt.virsh.numatune.positive_testing.set_numa_parameter.running_guest.change_nodeset.options.live.comma_list: SKIP
(16/60) type_specific.io-github-autotest-libvirt.virsh.numatune.positive_testing.set_numa_parameter.running_guest.change_nodeset.options.live.ranges: SKIP
(17/60) type_specific.io-github-autotest-libvirt.virsh.numatune.positive_testing.set_numa_parameter.running_guest.change_nodeset.options.live.excluding: SKIP
(18/60) type_specific.io-github-autotest-libvirt.virsh.numatune.positive_testing.set_numa_parameter.running_guest.change_nodeset.options.live.single: PASS (1.32 s)
(19/60) type_specific.io-github-autotest-libvirt.virsh.numatune.positive_testing.set_numa_parameter.running_guest.change_nodeset.options.current.comma_list: SKIP
(20/60) type_specific.io-github-autotest-libvirt.virsh.numatune.positive_testing.set_numa_parameter.running_guest.change_nodeset.options.current.ranges: SKIP
(21/60) type_specific.io-github-autotest-libvirt.virsh.numatune.positive_testing.set_numa_parameter.running_guest.change_nodeset.options.current.excluding: SKIP
(22/60) type_specific.io-github-autotest-libvirt.virsh.numatune.positive_testing.set_numa_parameter.running_guest.change_nodeset.options.current.single: PASS (1.41 s)
(23/60) type_specific.io-github-autotest-libvirt.virsh.numatune.positive_testing.get_numa_parameter.running_guest.options.none: PASS (1.30 s)
(24/60) type_specific.io-github-autotest-libvirt.virsh.numatune.positive_testing.get_numa_parameter.running_guest.options.live: PASS (1.28 s)
(25/60) type_specific.io-github-autotest-libvirt.virsh.numatune.positive_testing.get_numa_parameter.running_guest.options.current: PASS (1.29 s)
(26/60) type_specific.io-github-autotest-libvirt.virsh.numatune.positive_testing.get_numa_parameter.shutoff_guest.options.none: PASS (0.69 s)
(27/60) type_specific.io-github-autotest-libvirt.virsh.numatune.positive_testing.get_numa_parameter.shutoff_guest.options.config: PASS (0.69 s)
(28/60) type_specific.io-github-autotest-libvirt.virsh.numatune.positive_testing.get_numa_parameter.shutoff_guest.options.current: PASS (0.69 s)
(29/60) type_specific.io-github-autotest-libvirt.virsh.numatune.negative_testing.get_numa_parameter.running_guest.options.none: PASS (1.30 s)
(30/60) type_specific.io-github-autotest-libvirt.virsh.numatune.negative_testing.get_numa_parameter.running_guest.cgroup.stop: PASS (4.03 s)
(31/60) type_specific.io-github-autotest-libvirt.virsh.numatune.negative_testing.get_numa_parameter.shutoff_guest.options.none: PASS (0.70 s)
(32/60) type_specific.io-github-autotest-libvirt.virsh.numatune.negative_testing.get_numa_parameter.shutoff_guest.options.config: PASS (0.68 s)
(33/60) type_specific.io-github-autotest-libvirt.virsh.numatune.negative_testing.set_numa_parameter.running_guest.change_mode.options.config.preferred: PASS (1.29 s)
(34/60) type_specific.io-github-autotest-libvirt.virsh.numatune.negative_testing.set_numa_parameter.running_guest.change_mode.options.config.interleave: PASS (1.30 s)
(35/60) type_specific.io-github-autotest-libvirt.virsh.numatune.negative_testing.set_numa_parameter.running_guest.change_mode.options.config.strict: PASS (1.34 s)
(36/60) type_specific.io-github-autotest-libvirt.virsh.numatune.negative_testing.set_numa_parameter.running_guest.change_nodeset.options.live.comma_list: PASS (1.27 s)
(37/60) type_specific.io-github-autotest-libvirt.virsh.numatune.negative_testing.set_numa_parameter.running_guest.change_nodeset.options.live.ranges: PASS (1.31 s)
(38/60) type_specific.io-github-autotest-libvirt.virsh.numatune.negative_testing.set_numa_parameter.running_guest.change_nodeset.options.live.excluding: PASS (1.28 s)
(39/60) type_specific.io-github-autotest-libvirt.virsh.numatune.negative_testing.set_numa_parameter.running_guest.change_nodeset.options.live.single: PASS (1.29 s)
(40/60) type_specific.io-github-autotest-libvirt.virsh.numatune.negative_testing.set_numa_parameter.running_guest.change_nodeset.options.config.comma_list: PASS (1.31 s)
(41/60) type_specific.io-github-autotest-libvirt.virsh.numatune.negative_testing.set_numa_parameter.running_guest.change_nodeset.options.config.ranges: PASS (1.30 s)
(42/60) type_specific.io-github-autotest-libvirt.virsh.numatune.negative_testing.set_numa_parameter.running_guest.change_nodeset.options.config.excluding: PASS (1.33 s)
(43/60) type_specific.io-github-autotest-libvirt.virsh.numatune.negative_testing.set_numa_parameter.running_guest.change_nodeset.options.config.single: PASS (1.31 s)
(44/60) type_specific.io-github-autotest-libvirt.virsh.numatune.negative_testing.set_numa_parameter.running_guest.change_nodeset.options.current.comma_list: PASS (1.31 s)
(45/60) type_specific.io-github-autotest-libvirt.virsh.numatune.negative_testing.set_numa_parameter.running_guest.change_nodeset.options.current.ranges: PASS (1.28 s)
(46/60) type_specific.io-github-autotest-libvirt.virsh.numatune.negative_testing.set_numa_parameter.running_guest.change_nodeset.options.current.excluding: PASS (1.29 s)
(47/60) type_specific.io-github-autotest-libvirt.virsh.numatune.negative_testing.set_numa_parameter.running_guest.change_nodeset.options.current.single: PASS (1.29 s)
(48/60) type_specific.io-github-autotest-libvirt.virsh.numatune.negative_testing.set_numa_parameter.running_guest.cgroup.stop: PASS (1.30 s)
(49/60) type_specific.io-github-autotest-libvirt.virsh.numatune.negative_testing.set_numa_parameter.shutoff_guest.change_nodeset.options.live.comma_list: PASS (0.70 s)
(50/60) type_specific.io-github-autotest-libvirt.virsh.numatune.negative_testing.set_numa_parameter.shutoff_guest.change_nodeset.options.live.ranges: PASS (0.68 s)
(51/60) type_specific.io-github-autotest-libvirt.virsh.numatune.negative_testing.set_numa_parameter.shutoff_guest.change_nodeset.options.live.excluding: PASS (0.69 s)
(52/60) type_specific.io-github-autotest-libvirt.virsh.numatune.negative_testing.set_numa_parameter.shutoff_guest.change_nodeset.options.live.single: PASS (0.68 s)
(53/60) type_specific.io-github-autotest-libvirt.virsh.numatune.negative_testing.set_numa_parameter.shutoff_guest.change_nodeset.options.config.comma_list: PASS (0.72 s)
(54/60) type_specific.io-github-autotest-libvirt.virsh.numatune.negative_testing.set_numa_parameter.shutoff_guest.change_nodeset.options.config.ranges: PASS (0.72 s)
(55/60) type_specific.io-github-autotest-libvirt.virsh.numatune.negative_testing.set_numa_parameter.shutoff_guest.change_nodeset.options.config.excluding: PASS (0.76 s)
(56/60) type_specific.io-github-autotest-libvirt.virsh.numatune.negative_testing.set_numa_parameter.shutoff_guest.change_nodeset.options.config.single: PASS (0.69 s)
(57/60) type_specific.io-github-autotest-libvirt.virsh.numatune.negative_testing.set_numa_parameter.shutoff_guest.change_nodeset.options.current.comma_list: PASS (0.71 s)
(58/60) type_specific.io-github-autotest-libvirt.virsh.numatune.negative_testing.set_numa_parameter.shutoff_guest.change_nodeset.options.current.ranges: PASS (0.71 s)
(59/60) type_specific.io-github-autotest-libvirt.virsh.numatune.negative_testing.set_numa_parameter.shutoff_guest.change_nodeset.options.current.excluding: PASS (0.71 s)
(60/60) type_specific.io-github-autotest-libvirt.virsh.numatune.negative_testing.set_numa_parameter.shutoff_guest.change_nodeset.options.current.single: PASS (0.68 s)
TOTAL TIME: 69.94 s (01:09)
TESTS PASSED: 54
TESTS FAILED: 0
TESTS SKIPPED: 6
SUCCESS RATE: 100.00 %

**Comment by will-Do** [comment]
Okay, let's give a conclusion for this PR

**Comment by PandaWei** [comment]
@yangdongsheng  @will-Do 
This PR has last almost two months
If you haven't any other comments,
May i trouble you help me merge this?

**Comment by yangdongsheng** [comment]
Okey, as you pasted the results from your testing. it seems safe enough to me. Let's merge it. @will-Do , what's your opinion?

**Comment by PandaWei** [comment]
@will-Do 
what's your opinion about this PR?
It had got one ACK from @yangdongsheng

**Comment by will-Do** [comment]
@PandaWei 
Oh, I almost forgot this PR.
The codes and test results are good, I will merge it now.
Thanks.

**Pull request merged**